execution-engine/blessed-contracts/mint-token/src/lib.rs

```
@@ -0,0 +1,120 @@
+#![no_std]
+#![feature(alloc, cell_update)]
+
+#[macro_use]
+extern crate alloc;
+extern crate cl_std;
+
+mod capabilities;
+mod mint;
+
+use alloc::collections::BTreeMap;
+use alloc::string::String;
+use core::convert::TryInto;
+
+use capabilities::{ARef, RRef, RWRef, Readable};
+use cl_std::contract_api;
+use cl_std::key::Key;
+use cl_std::value::U512;
+use mint::Mint;
+
+type UAtom = RWRef<()>;
+type WeakUAtom = RRef<()>;
+
+struct CLMint;
+
+impl Mint<ARef<U512>, RWRef<U512>> for CLMint {
+    type PurseId = UAtom;
+    type DepOnlyId = WeakUAtom;
+
+    fn create(&self, balance: U512) -> Self::PurseId {
+        let purse = contract_api::new_uref(());
+        let balance = contract_api::new_uref(balance);
+
+        // Save purse -> balance association for future reference.
+        // TODO: do something more efficient than using strings as IDs
+        contract_api::add_uref(&purse_name(purse.0), &balance.into());
+
+        purse.try_into().unwrap()
```

Maybe this conversion should be tried before we call `add_uref`. If we can fail lets fail fast before we make more "expensive" calls.

---

The conversion will never fail though because we know that `new_uref` creates a `UPointer` with full permissions.